File: python/qc/wave_packet/bench.py
"""
Benchmark the NumPy wave packet step.

Outputs JSON: {"language":"python_numpy","example":"wave_packet","ns_per_op":...,"iterations":...}
"""

import time
import json
from physics import Sim, BARRIER_H_DEF

WARMUP = 200
ITERS  = 2_000

def main():
    sim = Sim(BARRIER_H_DEF)

    for _ in range(WARMUP):
        sim.step()
    sim.init_psi()

    t0 = time.perf_counter_ns()
    for _ in range(ITERS):
        sim.step()
    t1 = time.perf_counter_ns()

    ns = (t1 - t0) / ITERS
    print(json.dumps({
        "language":   "python_numpy",
        "example":    "wave_packet",
        "ns_per_op":  round(ns, 2),
        "iterations": ITERS,
    }))

if __name__ == "__main__":
    main()

File: rust/qc/wave_packet/src/bench.rs
//! Benchmark the split-operator TDSE step.
//!
//! Outputs JSON: {"language":"rust","example":"wave_packet","ns_per_op":...,"iterations":...}

mod physics;
use physics::Sim;
use std::hint::black_box;
use std::time::Instant;

const WARMUP: usize  = 200;
const ITERS:  usize  = 2_000;

fn main() {
    let mut sim = Sim::new(physics::BARRIER_H_DEF);

    // Warmup — let CPU reach steady-state clock and rustfft cache plans
    for _ in 0..WARMUP {
        black_box(&mut sim).step();
    }
    sim.init_psi();

    // Timed loop
    let start = Instant::now();
    for _ in 0..ITERS {
        black_box(&mut sim).step();
    }
    let elapsed = start.elapsed();

    let ns = elapsed.as_nanos() as f64 / ITERS as f64;
    println!(
        r#"{{"language":"rust","example":"wave_packet","ns_per_op":{:.2},"iterations":{}}}"#,
        ns, ITERS
    );
}
